File: 2023/day5/python/day5.py
# Preprocess input by removing all text
input = open("2023/day5/test.txt").read().strip().split('\n')

# part 1
seeds = input[0].split()
maps = []
c = -1
for l in input[1:]:
    if l == "":
        maps.append([]) 
        c += 1
        continue

    d,s,r = l.split()
    maps[c].append((int(d),int(s),int(r)))


res = []
for seed in seeds:
    seed = int(seed)
    for m in maps:
        for d,s,r in m:
            t = seed - s
            if t in range(r):
                seed = d + t
                break
    res.append(seed)

print("part1: ", min(res))

# part 2
it = iter(input[0].split())
seeds = [(x, next(it)) for x in it]

maps = []
for l in input[1:]:
    if l == "":
        maps.append([]) 
        c += 1
        continue

    d,s,r = l.split()
    maps[c].append((int(d),int(s),int(r)))

print(seeds)

File: 2023/day5/haskell/day5.hs
import Data.List
import Data.List.Split

getSingleMapping :: [Int] -> Int -> Maybe Int
getSingleMapping [d, s, r] v = if v >= d && v < d + r then Just $ s + (v - d) else Nothing
getSingleMapping _ _ = Nothing

-- map a value
getMapping :: [[Int]] -> Int -> Maybe Int
getMapping (m : ms) v = case getSingleMapping m v of
  Just o -> Just o
  Nothing -> getMapping ms v
getMapping [] v = Just v

-- has this value (after all mappings applied)a corresponding seed or not?
checkSeed :: [Int] -> Int -> Bool
checkSeed [s, r] v = v >= s && v < s + r
checkSeed _ _ = False

-- does a path exist for this value?
hasPath :: [[[Int]]] -> [[Int]] -> Int -> Bool
hasPath [] (s : seeds) v = checkSeed s v || hasPath [] seeds v
hasPath [] [] v = False
hasPath (m : maps) seeds v = case getMapping m v of
  Just o -> hasPath maps seeds o
  Nothing -> False

-- loop through locations beginning from low - we are looking for the lowest number
loopLocations :: [[[Int]]] -> [[Int]] -> Int -> Int
loopLocations maps seeds i =
  if hasPath maps seeds i then i else loopLocations maps seeds (i + 1)

main = do
  -- file must have the string lines removed and the first line split into a seed range entry per
  -- line
  input <- readFile "../input.txt"
  let r = map (map read . words) (lines input) :: [[Int]]
  let seeds : maps = splitWhen null r

  print seeds
  print $ reverse maps
  -- do some pivoting with bigger steps to find a rough number, then search by increasing by one
  let lowest = loopLocations (reverse maps) seeds 23738000

  print lowest
  return ()
